File: dev/js/PoolLayer.js
"use strict"

class PoolLayer {

    constructor (size, {stride, activation}={}) {

        if (size)   this.size = size
        if (stride) this.stride = stride

        if (activation!=undefined && activation!=false) {
            this.activation = typeof activation=="function" ? activation : NetMath[NetUtil.format(activation)].bind(this)
        } else {
            this.activation = false
        }
    }

    init () {}

    assignNext (layer) {
        this.nextLayer = layer
    }

    assignPrev (layer, layerIndex) {

        this.prevLayer = layer
        this.size = this.size || this.net.pool.size || 2
        this.stride = this.stride || this.net.pool.stride || this.size
        this.layerIndex = layerIndex

        let prevLayerOutWidth = layer.outMapSize

        switch (layer.constructor.name) {

            case "FCLayer":
                this.channels = this.net.channels
                prevLayerOutWidth = Math.max(Math.floor(Math.sqrt(layer.size/this.channels)), 1)
                break

            case "ConvLayer":
                this.channels = layer.size
                break

            case "PoolLayer":
                this.channels = layer.channels
                break
        }

        this.prevLayerOutWidth = prevLayerOutWidth
        this.outMapSize = (prevLayerOutWidth - this.size) / this.stride + 1
        this.inMapValuesCount = prevLayerOutWidth ** 2

        if (this.outMapSize%1 != 0) {
            throw new Error(`Misconfigured hyperparameters. Activation volume dimensions would be ${this.outMapSize} in pool layer at index ${layerIndex}`)
        }

        this.activations = [...new Array(this.channels)].map(channel => {
            return [...new Array(this.outMapSize)].map(row => [...new Array(this.outMapSize)].map(v => 0))
        })
        this.errors = [...new Array(this.channels)].map(channel => {
            return [...new Array(prevLayerOutWidth)].map(row => [...new Array(prevLayerOutWidth)].map(v => 0))
        })
        this.indeces = this.activations.map(channel => channel.map(row => row.map(v => [0,0])))
    }

    forward () {
        for (let channel=0; channel<this.channels; channel++) {

            NetMath.maxPool(this, channel)

            // Apply activations
            if (this.activation) {
                for (let row=0; row<this.outMapSize; row++) {
                    for (let col=0; col<this.outMapSize; col++) {
                        this.activations[channel][row][col] = this.activation(this.activations[channel][row][col], false, this.net)
                    }
                }
            }
        }
    }

    backward () {

        // Clear the existing error values, first
        for (let channel=0; channel<this.channels; channel++) {
            for (let row=0; row<this.errors[0].length; row++) {
                for (let col=0; col<this.errors[0].length; col++) {
                    this.errors[channel][row][col] = 0
                }
            }
        }

        if (this.nextLayer instanceof FCLayer) {

            for (let channel=0; channel<this.channels; channel++) {
                for (let row=0; row<this.outMapSize; row++) {
                    for (let col=0; col<this.outMapSize; col++) {

                        const rowI = this.indeces[channel][row][col][0] + row * this.stride
                        const colI = this.indeces[channel][row][col][1] + col * this.stride
                        const weightIndex = channel * this.outMapSize**2 + row * this.outMapSize + col

                        for (let neuron=0; neuron<this.nextLayer.neurons.length; neuron++) {
                            this.errors[channel][rowI][colI] += this.nextLayer.neurons[neuron].error
                                                                * this.nextLayer.neurons[neuron].weights[weightIndex]
                        }
                    }
                }
            }

        } else if (this.nextLayer instanceof ConvLayer) {

            for (let channel=0; channel<this.channels; channel++) {

                const errs = []

                for (let col=0; col<this.outMapSize; col++) {
                    errs[col] = 0
                }

                // Convolve on the error map
                NetUtil.buildConvErrorMap(this.nextLayer, errs, channel)

                for (let row=0; row<this.outMapSize; row++) {
                    for (let col=0; col<this.outMapSize; col++) {

                        const rowI = this.indeces[channel][row][col][0] + row * this.stride
                        const colI = this.indeces[channel][row][col][1] + col * this.stride

                        this.errors[channel][rowI][colI] += errs[row][col]
                    }
                }
            }

        } else {

            for (let channel=0; channel<this.channels; channel++) {
                for (let row=0; row<this.outMapSize; row++) {
                    for (let col=0; col<this.outMapSize; col++) {

                        const rowI = this.indeces[channel][row][col][0] + row * this.stride
                        const colI = this.indeces[channel][row][col][1] + col * this.stride

                        this.errors[channel][rowI][colI] += this.nextLayer.errors[channel][row][col]
                    }
                }
            }
        }

        // Apply derivatives
        if (this.activation) {
            for (let channel=0; channel<this.channels; channel++) {

                for (let row=0; row<this.indeces[channel].length; row++) {
                    for (let col=0; col<this.indeces[channel].length; col++) {

                        const rowI = this.indeces[channel][row][col][0] + row * this.stride
                        const colI = this.indeces[channel][row][col][1] + col * this.stride

                        this.errors[channel][rowI][colI] *= this.activation(this.errors[channel][rowI][colI], true, this.net)
                    }
                }
            }
        }
    }

    resetDeltaWeights () {}

    applyDeltaWeights () {}

    toJSON () {return {}}

    fromJSON () {}
}

typeof window=="undefined" && (exports.PoolLayer = PoolLayer)

File: dev/cpp/PoolLayer.cpp

PoolLayer::PoolLayer (int netI, int s) : Layer(netI, s) {
    netInstance = netI;
    size = s;
    type = "Pool";
    hasActivation = false;
}

PoolLayer::~PoolLayer (void) {}

void PoolLayer::assignNext (Layer* l) {
    nextLayer = l;
}

void PoolLayer::assignPrev (Layer* l) {
    prevLayer = l;
}

void PoolLayer::init (int layerIndex) {
    prevLayerOutWidth = sqrt(inMapValuesCount);
    activations = NetUtil::createVolume(channels, outMapSize, outMapSize, 0.0);
    errors = NetUtil::createVolume(channels, prevLayerOutWidth, prevLayerOutWidth, 0.0);
    std::vector<int> emptyIndex = {0,0};
    indeces = NetUtil::createVolume(channels, outMapSize, outMapSize, emptyIndex);
}

void PoolLayer::forward (void) {

    for (int channel=0; channel<channels; channel++) {

        NetMath::maxPool(this, channel);

        // Apply activations
        if (hasActivation) {
            for (int r=0; r<outMapSize; r++) {
                for (int v=0; v<outMapSize; v++) {
                    activations[channel][r][v] = activationP(activations[channel][r][v], false, Network::getInstance(netInstance));
                }
            }
        }
    }
}

void PoolLayer::backward (void) {

    // Clear the existing error values, first
    for (int c=0; c<channels; c++) {
        for (int r=0; r<prevLayerOutWidth; r++) {
            for (int v=0; v<prevLayerOutWidth; v++) {
                errors[c][r][v] = 0;
            }
        }
    }

    if (nextLayer->type=="FC") {

        for (int c=0; c<channels; c++) {
            for (int r=0; r<outMapSize; r++) {
                for (int v=0; v<outMapSize; v++) {

                    int rowI = indeces[c][r][v][0] + r * stride;
                    int colI = indeces[c][r][v][1] + v * stride;
                    int weightI = c * outMapSize*outMapSize + r * outMapSize + v;

                    for (int n=0; n<nextLayer->neurons.size(); n++) {
                        errors[c][rowI][colI] += nextLayer->neurons[n]->error * nextLayer->weights[n][weightI];
                    }
                }
            }
        }

    } else if (nextLayer->type=="Conv") {

        for (int c=0; c<channels; c++) {

            // Convolve on the error map
            std::vector<std::vector<double> > errs = NetUtil::buildConvErrorMap(outMapSize+nextLayer->zeroPadding*2, nextLayer, c);

            for (int r=0; r<outMapSize; r++) {
                for (int v=0; v<outMapSize; v++) {
                    int rowI = indeces[c][r][v][0] + r * stride;
                    int colI = indeces[c][r][v][1] + v * stride;

                    errors[c][rowI][colI] += errs[r][v];
                }
            }
        }

    } else {

        for (int c=0; c<channels; c++) {
            for (int r=0; r<outMapSize; r++) {
                for (int v=0; v<outMapSize; v++) {

                    int rowI = indeces[c][r][v][0] + r * stride;
                    int colI = indeces[c][r][v][1] + v * stride;

                    errors[c][rowI][colI] += nextLayer->errors[c][r][v];
                }
            }
        }
    }

    // Apply derivatives
    if (hasActivation) {
        for (int c=0; c<channels; c++) {
            for (int r=0; r<outMapSize; r++) {
                for (int v=0; v<outMapSize; v++) {

                    int rowI = indeces[c][r][v][0] + r * stride;
                    int colI = indeces[c][r][v][1] + v * stride;

                    errors[c][rowI][colI] *= activationP(errors[c][rowI][colI], true, Network::getInstance(netInstance));
                }
            }
        }
    }
}
